Rewrite documentation - Main Concepts - Permissions (#1845)
kuzzleio/kuzzle: doc/2/guides/main-concepts/3-permissions/index.md

---

## Initial Setup

When you run your application for the first time there is no administrator account and anonymous users (i.e. unauthenticated users) can execute any API action.

To secure your application you will need to create an administrator account by using the [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) API action.

::: info
The [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) action creates a user attached to the `admin` profile, which uses the `admin` role, giving access to all API actions.  
The `reset` option allows to restrict `anonymous` default rights in the same time.
:::

```bash
$ kourou security:createFirstAdmin '{
  credentials: {
    local: {
      username: "admin",
      password: "password"
    }
  }
}' -a reset=true
```

Once the administrator account is created, and anonymous access rights are removed, you can properly secure your installation.  

::: info
You can then use the [Admin Console](http://console.kuzzle.io) or Kuzzle API to create new users and assign them permissions.
:::
---

## Roles

> Les roles permettent de définir une liste d'action d'API qui seront authorisées.

Author: ```suggestion
A role details a set of allowed API actions.
```

```

In the above role definition, anonymous users can perform the following actions [auth:login](/core/2/api/controllers/auth/login), [auth:checkToken](/core/2/api/controllers/auth/check-token), [auth:getCurrentUser](/core/2/api/controllers/auth/get-current-user) and [auth:getMyRights](/core/2/api/controllers/auth/get-my-rights).

::: info
Default roles and profiles used once the administrator account is created with the [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) action can be configured under the [security.default](/core/2/guides/advanced/8-configuration) configuration key.

> :::

Author: This guide is not a quick guide but it should rather offer a complete overview of the Permission system feature

> :::

Reviewer: oh, ok

}
```

::: info

> Role restriction on indexes and collections can be used to **easily build multi-tenant application** were different groups of users cannot see each other data.

Reviewer 2: ```suggestion
Role restriction on indexes and collections can be used to **easily build multi-tenant application** where different groups of users cannot see each other data.
```

:::

### Rate limit

<SinceBadge version="2.1.0" />

The rate limit parameter controls how many API requests a user can send, **per second and per node**. Further requests made by a user that exceed the limit are rejected with a `429 Too Many Requests` error.

::: info
If no rate limit is defined, or if it is set to 0, then no limit is applied.
If a user has several profiles with rate limits, the most permissive limit applies.
:::

::: warning

> Since unauthenticated users share the same user identifier, a rate limit set on the `anonymous` profile is applied to **all anonymous requests cumulated**, per second and per node. Except for the `auth:login` route, which is statically controlled in Kuzzle's configuration files.

Reviewer: (suggestion/open to debate)

```suggestion
Since unauthenticated users share the same user identifier, a rate limit set on the `anonymous` profile is applied to **all anonymous requests cumulated**, per second and per node. Except for the `auth:login` route, which is handled separately to allow users to escalate their priviledges even if Kuzzle gets overloaded by anonymous requests.
```

Author: I'm not fond of the `escalate their privileges` part because privilege escalation could be a scary word :sweat_smile:

Reviewer: :+1: 

I've update my suggestion.

:::

Example:

```js
{
  "rateLimit": 20,
  "policies": [ /* ...role policies, see below ... */ ]
}
```

### Associated API actions

 - [security:createOrReplaceProfile](/core/2/api/controllers/security/create-or-replace-profile/)
 - [security:createProfile](/core/2/api/controllers/security/create-profile/)
 - [security:deleteProfile](/core/2/api/controllers/security/delete-profile/)
 - [security:getProfile](/core/2/api/controllers/security/get-profile/)
 - [security:getProfileMapping](/core/2/api/controllers/security/get-profile-mapping/)
 - [security:getProfileRights](/core/2/api/controllers/security/get-profile-rights/)
 - [security:mDeleteProfiles](/core/2/api/controllers/security/m-delete-profiles/)
 - [security:mGetProfiles](/core/2/api/controllers/security/m-get-profiles/)
 - [security:scrollProfiles](/core/2/api/controllers/security/scroll-profiles/)
 - [security:searchProfiles](/core/2/api/controllers/security/search-profiles/)
 - [security:updateProfile](/core/2/api/controllers/security/update-profile/)
 - [security:updateProfileMapping](/core/2/api/controllers/security/update-profile-mapping/)

---

## Users

> Users can be assigned multiple profiles. These profiles will give them access to API actions via roles and their access will be limited by any restrictions defined in the profiles.

Reviewer: ```suggestion
Users can have multiple profiles assigned to them. These profiles will give them access to API actions via roles, and their access will be limited by any restrictions defined in the profiles.
```

Users are internal documents stored by Kuzzle. They contain two properties:
 - `content`: profiles list and custom content

>  - `credentials`: available authentication credentials

Reviewer: There are no credentials information stored in a user documents. I think that not even the list of known credentials for that user is stored. 
Only authentication plugins have that kind of information, there are strict separations between Kuzzle users and their credentials to prevent security issues.

::: warning
The information contained in the `credentials` property are never returned and can only be accessed by the plugin that added the associated [authentication strategy](/core/2/guides/write-plugins/3-integrate-authentication-strategy).
More information about [Authentication](/core/2/some-link).
:::

### Profiles list

The list of profiles assigned to a user is contained in the `content.profileIds` property.  
This property is a list of profile identifiers:

```js
{
  "content": {
    "profileIds": ["publisher", "reader"]
  }
}
```

::: info
Users must be assigned to at least one profile.
:::

### Custom content

It is also possible to store custom data in the user object.  

Those data must be stored in the `content` property alongside the profile list:

```js
{
  "content": {
    "profileIds": ["publisher", "reader"],

    // custom data
    "firstname": "Clément",
    "lastname": "M'bileau"
  }
}
```

::: info

> As any other collection, the `users` collection has an associated mapping who can be edited using the [security:updateUserMapping](/core/2/api/controllers/security/update-user-mapping/) API action.

Reviewer: ```suggestion
As any other collection, the `users` collection has an associated mapping that can be edited using the [security:updateUserMapping](/core/2/api/controllers/security/update-user-mapping/) API action.
```

## Writing complex or dynamic permission rules

So far, we've seen how to set permissions to API action, using user roles and profiles.

But this is rarely enough to secure an application, as it's commonplace to **reject queries or data depending on business rules**.

For instance, suppose you have a chat application and you want the users to only be able to edit & delete their own messages: this type of rules cannot be expressed as a simple boolean.

There are multiple ways of adding a business logic layer on top of the standard Kuzzle security one:

* With a [Pipe](/core/2/guides/develop-on-kuzzle/3-event-system#pipes), you can listen to one or multiple [API events](/core/2/some-link), and decide whether you accept a query or document according to your business rules (you can see an example on [Github](https://github.com/kuzzleio/kuzzle-plugin-sample-custom-policies))
* If all you need is to make sure that submitted documents follow a strict set of formatting rules, you can add [document validators](/core/2/guides/advanced/9-data-validation)

::: info
More information about dynamic rules with pipes: [Event System](/core/2/some-link)
::: 

## Load permissions

> admin controller

Reviewer 2: what is this?

Author: WIP :sweat_smile:

Reviewer 2: Oh, OK. We should use a comment here, no?

Reviewer 2: (not important)

Author: I added the section :+1:
